From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Loki Rautio <lokirautio@gmail.com>
Date: Thu, 19 Jun 2025 15:27:46 -0500
Subject: [PATCH] Allow force filling the trades during TradeAcquire

ensure difference

diff --git a/net/minecraft/world/entity/npc/AbstractVillager.java b/net/minecraft/world/entity/npc/AbstractVillager.java
index e22749e55dcab2a023a3010c34b1cd45fe5aa895..652211531ba3ba718992d6943602c0c09ec861a7 100644
--- a/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -272,9 +272,10 @@ public abstract class AbstractVillager extends AgeableMob implements InventoryCa
 
     protected void addOffersFromItemListings(MerchantOffers givenMerchantOffers, VillagerTrades.ItemListing[] newTrades, int maxNumbers) {
         ArrayList<VillagerTrades.ItemListing> list = Lists.newArrayList(newTrades);
+        int originalTradeAmount = givenMerchantOffers.size();
         int i = 0;
 
-        while (i < maxNumbers && !list.isEmpty()) {
+        while (i < maxNumbers && (givenMerchantOffers.size() - originalTradeAmount) < maxNumbers && !list.isEmpty()) { // SMPOnline - Allow forcing trades
             MerchantOffer offer = list.remove(this.random.nextInt(list.size())).getOffer(this, this.random);
             if (offer != null) {
                 // CraftBukkit start
