From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Loki Rautio <lokirautio@gmail.com>
Date: Sat, 5 Jul 2025 21:56:40 -0500
Subject: [PATCH] Player lag compensation


diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index e25a2bfec795a3157d3dab4b74056e2674b0f394..9e59061b42dcf78cd9b4d6dbd9e7dff3a1fcfd38 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -7,6 +7,8 @@ import com.google.common.math.IntMath;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Either;
 import com.mojang.logging.LogUtils;
+
+import java.time.Instant;
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
@@ -204,6 +206,7 @@ public abstract class Player extends LivingEntity {
     // CraftBukkit start
     public boolean fauxSleeping;
     public int oldLevel = -1;
+    public Instant lastPlayTimeTick = Instant.now();
 
     @Override
     public org.bukkit.craftbukkit.entity.CraftHumanEntity getBukkitEntity() {
@@ -262,17 +265,21 @@ public abstract class Player extends LivingEntity {
 
     @Override
     public void tick() {
+        long millisSinceLastTick = Instant.now().toEpochMilli() - this.lastPlayTimeTick.toEpochMilli();
+        this.lastPlayTimeTick = Instant.now();
+        int ticksToCatchup = (int) Math.max(1, Math.round(millisSinceLastTick / 50d));
+
         this.noPhysics = this.isSpectator();
         if (this.isSpectator() || this.isPassenger()) {
             this.setOnGround(false);
         }
 
         if (this.takeXpDelay > 0) {
-            this.takeXpDelay--;
+            this.takeXpDelay-=ticksToCatchup;
         }
 
         if (this.isSleeping()) {
-            this.sleepCounter++;
+            this.sleepCounter+=ticksToCatchup;
             // Paper start - Add PlayerDeepSleepEvent
             if (this.sleepCounter == SLEEP_DURATION) {
                 if (!new io.papermc.paper.event.player.PlayerDeepSleepEvent((org.bukkit.entity.Player) getBukkitEntity()).callEvent()) {
@@ -288,7 +295,7 @@ public abstract class Player extends LivingEntity {
                 this.stopSleepInBed(false, true);
             }
         } else if (this.sleepCounter > 0) {
-            this.sleepCounter++;
+            this.sleepCounter+=ticksToCatchup;
             if (this.sleepCounter >= 110) {
                 this.sleepCounter = 0;
             }
@@ -303,19 +310,21 @@ public abstract class Player extends LivingEntity {
 
         this.moveCloak();
         if (this instanceof ServerPlayer serverPlayer) {
-            this.foodData.tick(serverPlayer);
-            this.awardStat(Stats.PLAY_TIME);
-            this.awardStat(Stats.TOTAL_WORLD_TIME);
-            if (this.isAlive()) {
-                this.awardStat(Stats.TIME_SINCE_DEATH);
-            }
+            for(int i = 0; i < ticksToCatchup; i++) {
+                this.foodData.tick(serverPlayer);
+                this.awardStat(Stats.PLAY_TIME);
+                this.awardStat(Stats.TOTAL_WORLD_TIME);
+                if (this.isAlive()) {
+                    this.awardStat(Stats.TIME_SINCE_DEATH);
+                }
 
-            if (this.isDiscrete()) {
-                this.awardStat(Stats.CROUCH_TIME);
-            }
+                if (this.isDiscrete()) {
+                    this.awardStat(Stats.CROUCH_TIME);
+                }
 
-            if (!this.isSleeping()) {
-                this.awardStat(Stats.TIME_SINCE_REST);
+                if (!this.isSleeping()) {
+                    this.awardStat(Stats.TIME_SINCE_REST);
+                }
             }
         }
 
@@ -340,10 +349,12 @@ public abstract class Player extends LivingEntity {
             this.turtleHelmetTick();
         }
 
-        this.cooldowns.tick();
+        for(int j = 0; j < ticksToCatchup; j++) {
+            this.cooldowns.tick();
+        }
         this.updatePlayerPose();
         if (this.currentImpulseContextResetGraceTime > 0) {
-            this.currentImpulseContextResetGraceTime--;
+            this.currentImpulseContextResetGraceTime-=ticksToCatchup;
         }
     }
 
