From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Loki Rautio <lokirautio@gmail.com>
Date: Sat, 28 Dec 2024 18:11:36 -0600
Subject: [PATCH] Add linear XP config entries


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 4ed27c10b432ceebf4447ab8007bc3a1be09a06e..8d32766484b01a26d75331de59da70a42c2c7b82 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -344,6 +344,8 @@ public class GlobalConfiguration extends ConfigurationPart {
         public IntOr.Default compressionLevel = IntOr.Default.USE_DEFAULT;
         @Comment("Defines the leniency distance added on the server to the interaction range of a player when validating interact packets.")
         public DoubleOr.Default clientInteractionLeniencyDistance = DoubleOr.Default.USE_DEFAULT;
+        public boolean useLinearXpFormula = false;
+        public int linearXpPerLevel = 17;
     }
 
     public BlockUpdates blockUpdates;
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index f2817f1c29431ef2c4a45dc9ef90f06d4982f7c9..5c4cf9382bfb1ecbeeefd3fa84fd2d6c8ffa1fc4 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1847,6 +1847,9 @@ public abstract class Player extends LivingEntity {
     }
 
     public int getXpNeededForNextLevel() {
+        if(io.papermc.paper.configuration.GlobalConfiguration.get().misc.useLinearXpFormula){
+            return io.papermc.paper.configuration.GlobalConfiguration.get().misc.linearXpPerLevel;
+        }
         return this.experienceLevel >= 30 ? 112 + (this.experienceLevel - 30) * 9 : (this.experienceLevel >= 15 ? 37 + (this.experienceLevel - 15) * 5 : 7 + this.experienceLevel * 2); // Paper - diff on change; calculateTotalExperiencePoints
     }
     // Paper start - send while respecting visibility
