From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Loki Rautio <lokirautio@gmail.com>
Date: Sat, 28 Dec 2024 21:59:24 -0600
Subject: [PATCH] Don't count trades not added towards addOffers


diff --git a/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java b/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java
index a1dee26ed47a423fbad57b9b2ec98b4f58b9470b..653bb46bc45fd21952ddeb8e035bd4e105f5078f 100644
--- a/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -299,9 +299,10 @@ public abstract class AbstractVillager extends AgeableMob implements InventoryCa
                     if (craftMerchantRecipe.getIngredients().isEmpty()) return;
                     recipeList.add(craftMerchantRecipe.toMinecraft());
                     // Paper end - Fix crash from invalid ingredient list
+                    ++j; // SMPOnline - Fix no trade behavior before exhausting the pool
                 }
                 // CraftBukkit end
-                ++j;
+
             }
         }
 
